### DISCUSSION TITLE
Segmentation fault on pre-split mesh

### DISCUSSION THREAD

**Discussion by heinono1** [idaholab/moose#24410] [Q&A Meshing]
Hi,

I typically use pre-split meshes as (just like in  #23713) my mesh is large and I cannot replicate (will run out of RAM). Pre-splitting has worked well with previous installations but after I updated and rebuilt MOOSE on May 3, including Libmesh and PETSc, I get errors: 
```
--------------------------------------------------------------------------
Primary job  terminated normally, but 1 process returned
a non-zero exit code. Per user-direction, the job has been aborted.
--------------------------------------------------------------------------
--------------------------------------------------------------------------
mpirun noticed that process rank 57 with PID 1637367 on node node05 exited on signal 11 (Segmentation fault).
--------------------------------------------------------------------------
```
This is uniquely related to pre-splitting (I have some nodes on which I can use replicated mesh and then this error does not happen) and uniquely related to the update on May 3. 

Any ideas of what is going on?

**Comment by GiudGiud** [comment]
Hello

Do you use multiapps?
From when was your last working version? so we can narrow down which PR changed it
Do you see anything helpful in debug mode? (just compiled in debug mode first, but running in a debugger could get us even more info)
Guillaume

**Reply by heinono1** [reply]
Hi Guillaume. No multiapps - just usual stuff and a UserObject. Haven't run debug mode - I'll see if I get around to it (busy cranking out results so may not have time for a while to dig into debug). 

I pulled the MOOSE, Libmesh, and PETSc (updating Libmesh and PETSc using the MOOSE scripts) on May 3. Here is the Framework information:
Framework Information:
MOOSE Version:           git commit 2da0ccfe5f on 2023-04-11
LibMesh Version:         80494219d2293e467392189814d30330c4f87a25
PETSc Version:           3.16.6
SLEPc Version:           3.16.2

**Reply by GiudGiud** [reply]
do you have the hash for the working version? Presumably just the previous one you used

**Reply by GiudGiud** [reply]
it will be in any simulation log

**Comment by heinono1** [comment]
It looks like this is related to the number of processes (MPI tasks). When I split 96 ways (96 cores in the node) I get the seg fault, but when I split 64 ways, things seem to work. I tried to split 96 ways using 8 or 16 tasks so as not to overwhelm RAM during splitting so I don't think that I ran out of RAM - looks like something else going on.

**Reply by GiudGiud** [reply]
We've changed a lot of things in how often we prepare the mesh (search for elem neighbors, delete remote elements, repartition) to do it as little as possible

This is probably tied. Can you share your mesh? or can you make a MWE?

**Reply by heinono1** [reply]
I am afraid I cannot share the mesh (IP). What's MWE?

**Reply by GiudGiud** [reply]
minimal (non) working example

If you can reproduce this error with a MOOSE mesh or anything else you can share

**Reply by heinono1** [reply]
I'll see if I can mock up some MWE. Isn't the hash MOOSE Version: git commit https://github.com/idaholab/moose/commit/2da0ccfe5f6ed2b2e00abf13acb1ca6b09458916 on 2023-04-11? Is it something else?

**Reply by GiudGiud** [reply]
That's for the working version then?

**Reply by heinono1** [reply]
That's the current version I am running that sometimes splits successfully, sometimes not.

**Reply by lindsayad** [reply]
> We've changed a lot of things in how often we prepare the mesh (search for elem neighbors, delete remote elements, repartition) to do it as little as possible

I think the changes we merged ensure that we prepare the mesh at least as much as before and sometimes more. The "more" preparing has revealed some buggy behavior both in MOOSE and in libMesh that we've worked quickly to fix. I know there is at least one outstanding issue on the libMesh side that @roystgnr is working on. Perhaps he can comment on how often he thinks that might crop up in real user cases

**Reply by roystgnr** [reply]
The only thing that I know we're not doing right yet is making sure that an already-constructed `PointLocator` gets reconstructed after mesh node positions are changed in the middle of a chain of mesh generators.  That's not going to be an easy case to hit, though; you'd need a mesh generator that uses a point locator, whose mesh is then fed to a generator that moves points, whose mesh is then fed to another generator that uses a point locator ... and even then you'd need the final generator to not do any kind of null-checking on the element returned by the locator.  That combination is probably not what's happening here.
